File: code/backend/src/application/services/quarto-service.ts
import { Quarto } from '../../domain/entities/Quarto';
import { IQuartoRepository } from '../repositories/quarto-repository';
import { CreateQuartoDTO, UpdateQuartoDTO, QuartoResponseDTO } from '../dtos/quarto-dto';
import { Cama, StatusQuarto, TipoCama } from '../../domain/types';
import { randomUUID } from 'crypto';
import { QuartoMapper } from '../mappers/quarto-mapper';

/**
 * Service Layer (Application Service):
 * Responsável por orquestrar os Casos de Uso do módulo de Quartos.
 * Decisão de Design:
 * - Não contém regras de negócio de domínio (essas ficam na Entity 'Quarto').
 * - Atua como uma fachada para a UI/API, convertendo DTOs em Entidades e vice-versa.
 * - Gerencia transações (se houvesse) e persistência via Repositório.
 */
export class QuartoService {
    // Injeção de Dependência (DIP): Dependemos da abstração (Interface), não da implementação concreta do Repositório.
    // Isso facilita testes unitários e troca de banco de dados.
    constructor(private quartoRepository: IQuartoRepository) {}

    async cadastrarQuarto(input: CreateQuartoDTO): Promise<QuartoResponseDTO> {
        // Validação de unicidade (Regra de Aplicação)
        const existente = await this.quartoRepository.findByNumero(input.numero);
        if (existente) {
            throw new Error(`Quarto com número ${input.numero} já existe.`);
        }

        // Refatorado: Lógica de geração extraída e simplificada
        const camas = this.gerarCamas(input.camas);

        // Factory/Builder implícito: Construção da entidade com estado inicial válido.
        const novoQuarto = new Quarto(
            randomUUID(),
            input.numero,
            input.capacidade,
            input.tipo,
            input.precoDiaria,
            StatusQuarto.LIVRE,
            camas,
            []
        );

        await this.quartoRepository.save(novoQuarto);

        // Uso de Mapper (SRP): Separa a responsabilidade de formatação de resposta da lógica de negócio.
        return QuartoMapper.toDTO(novoQuarto);
    }

    async editarQuarto(id: string, input: UpdateQuartoDTO): Promise<QuartoResponseDTO> {
        const quarto = await this.quartoRepository.findById(id);
        if (!quarto) {
            throw new Error(`Quarto com ID ${id} não encontrado.`);
        }

        // Refatorado: Reuso da lógica consistente
        // Transformação de Input DTO -> Value Objects / Entidades
        const novasCamas = input.camas ? this.gerarCamas(input.camas) : undefined;

        // Delegação para o Domínio: A entidade sabe como manter seus invariantes ao atualizar dados.
        quarto.atualizarDados(
            input.numero,
            input.capacidade,
            input.tipo,
            input.precoDiaria,
            novasCamas,
            undefined
        );

        await this.quartoRepository.update(quarto);

        return QuartoMapper.toDTO(quarto);
    }

    async listarQuartos(): Promise<QuartoResponseDTO[]> {
        const quartos = await this.quartoRepository.findAll();
        // Projeção eficiente de dados para o cliente.
        return quartos.map(q => QuartoMapper.toDTO(q));
    }

    /**
     * Helper Privado: Responsável por transformar a entrada simplificada (Tipo + Qtd)
     * em Entidades de Cama com identidade única.
     */
    private gerarCamas(camasInput?: { tipo: TipoCama; quantidade: number }[]): Cama[] {
        if (!camasInput) return [];

        return camasInput.flatMap(c =>
            Array.from({ length: c.quantidade }, () => ({
                id: randomUUID(),
                tipo: c.tipo
            }))
        );
    }
}

File: code/stack-2/backend/src/main/java/com/hotel/service/QuartoService.java
package com.hotel.service;

import com.hotel.dto.QuartoDTO;
import com.hotel.model.Quarto;
import com.hotel.repository.IQuartoRepository;
import org.springframework.stereotype.Service;
import java.util.List;

/**
 * Serviço responsável pela lógica de negócio da Gestão de Quartos.
 * Aplica o Princípio da Responsabilidade Única (SRP) ao focar apenas em regras de negócio,
 * delegando a persistência para a abstração do repositório.
 */
@Service
public class QuartoService {
    
    // Injeção via Interface (DIP): O serviço não depende de uma implementação concreta.
    private final IQuartoRepository repository;

    public QuartoService(IQuartoRepository repository) {
        this.repository = repository;
    }

    /**
     * Cadastra um novo quarto validando a unicidade do número.
     * @param dto Dados de entrada validados pela camada de Controller.
     */
    public Quarto cadastrar(QuartoDTO dto) {
        repository.findByNumero(dto.numero())
            .ifPresent(q -> {
                throw new RuntimeException("Regra de Negócio: Quarto " + dto.numero() + " já existe.");
            });

        Quarto quarto = mapToEntity(dto, new Quarto());
        return repository.save(quarto);
    }

    /**
     * Edita informações de um quarto existente.
     * @param numero Identificador único do quarto.
     * @param dto Novos dados.
     */
    public Quarto editar(String numero, QuartoDTO dto) {
        Quarto quarto = repository.findByNumero(numero)
            .orElseThrow(() -> new RuntimeException("Erro: Quarto não encontrado para edição."));
        
        mapToEntity(dto, quarto);
        return repository.save(quarto);
    }

    /**
     * Retorna todos os quartos cadastrados.
     */
    public List<Quarto> listarTodos() {
        return repository.findAll();
    }

    /**
     * Método auxiliar de mapeamento (Data Mapper Pattern simplificado).
     * Isola a lógica de transformação de DTO para Entidade, mantendo o código limpo.
     */
    private Quarto mapToEntity(QuartoDTO dto, Quarto entity) {
        entity.setNumero(dto.numero());
        entity.setCapacidade(dto.capacidade());
        entity.setTipo(dto.tipo());
        entity.setPrecoDiaria(dto.precoDiaria());
        entity.setTemFrigobar(dto.temFrigobar());
        entity.setTemCafeIncluso(dto.temCafeIncluso());
        entity.setTemArCondicionado(dto.temArCondicionado());
        entity.setTemTV(dto.temTV());
        entity.setTiposCama(dto.tiposCama());
        return entity;
    }
}
